ops/testing.py

```diff
@@ -2937,17 +2937,45 @@ def _render_services(self) -> Dict[str, pebble.Service]:
             layer = self._layers[key]
             for name, service in layer.services.items():
                 # TODO: (jam) 2021-04-07 have a way to merge existing services
+                # FIXME: see https://github.com/canonical/operator/issues/1112
                 services[name] = service
         return services
 
+    def checks(self) -> Dict[str, pebble.Check]:
+        """The checks in the current plan."""
+        services: Dict[str, pebble.Check] = {}
+        for key in sorted(self._layers.keys()):
+            layer = self._layers[key]
+            for name, check in layer.checks.items():
+                services[name] = check
+        return services
+
+    def log_targets(self) -> Dict[str, pebble.LogTarget]:
+        """The log targets in the current plan."""
+        services: Dict[str, pebble.LogTarget] = {}
+        for key in sorted(self._layers.keys()):
+            layer = self._layers[key]
+            for name, log_target in layer.log_targets.items():
+                services[name] = log_target
+        return services
+
     def get_plan(self) -> pebble.Plan:
         self._check_connection()
-        plan = pebble.Plan('{}')
+        plan = pebble.Plan("")
+
+        # FIXME: this horrible mutation
         services = self._render_services()
-        if not services:
-            return plan
-        for name in sorted(services.keys()):
-            plan.services[name] = services[name]
+        if services:
+            plan._services = services
+
+        checks = self.checks()
+        if checks:
+            plan._checks = checks
+
+        log_targets = self.log_targets()
+        if log_targets:
+            plan._log_targets = log_targets
+
```

Comment:
It might trigger your mutation spidey sense even more :-), but I think we should both simplify and avoid use of private `Plan` fields here by doing:

```
plan.services.update(self._render_services())
plan.checks.update(self._render_checks())
plan.log_targets.update(self._render_log_targets())
```

Reply:
well at least it's terse